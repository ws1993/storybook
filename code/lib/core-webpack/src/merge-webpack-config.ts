import type {
  WebpackConfiguration as Configuration,
  ModuleConfig,
  ResolveConfig,
  RulesConfig,
} from './types';

function mergePluginsField(
  defaultPlugins: Required<Configuration>['plugins'] = [],
  customPlugins: Required<Configuration>['plugins'] = []
): Required<Configuration>['plugins'] {
  return [...defaultPlugins, ...customPlugins];
}

function mergeRulesField(
  defaultRules: RulesConfig[] = [],
  customRules: RulesConfig[] = []
): ModuleConfig['rules'] {
  return [...defaultRules, ...customRules];
}

function mergeExtensionsField(
  { extensions: defaultExtensions = [] }: ResolveConfig,
  { extensions: customExtensions = [] }: ResolveConfig
): ResolveConfig['extensions'] {
  return [...defaultExtensions, ...customExtensions];
}

function mergeAliasField(
  { alias: defaultAlias = {} }: ResolveConfig,
  { alias: customAlias = {} }: ResolveConfig
): ResolveConfig['alias'] {
  return {
    ...defaultAlias,
    ...customAlias,
  };
}

function mergeModuleField(a: ModuleConfig, b: ModuleConfig): ModuleConfig {
  return {
    ...a,
    ...b,
    rules: mergeRulesField(a.rules || [], b.rules || []),
  };
}

function mergeResolveField(
  { resolve: defaultResolve = {} }: Configuration,
  { resolve: customResolve = {} }: Configuration
): ResolveConfig {
  return {
    ...defaultResolve,
    ...customResolve,
    alias: mergeAliasField(defaultResolve, customResolve),
    extensions: mergeExtensionsField(defaultResolve, customResolve),
  };
}

function mergeOptimizationField(
  { optimization: defaultOptimization = {} }: Configuration,
  { optimization: customOptimization = {} }: Configuration
): Configuration['optimization'] {
  return {
    ...defaultOptimization,
    ...customOptimization,
  };
}

export function mergeConfigs(config: Configuration, customConfig: Configuration): Configuration {
  return {
    // We'll always load our configurations after the custom config.
    // So, we'll always load the stuff we need.
    ...customConfig,
    ...config,
    devtool: customConfig.devtool || config.devtool,
    plugins: mergePluginsField(config.plugins, customConfig.plugins),
    module: mergeModuleField(config.module || {}, customConfig.module || {}),
    resolve: mergeResolveField(config, customConfig),
    optimization: mergeOptimizationField(config, customConfig),
  };
}
