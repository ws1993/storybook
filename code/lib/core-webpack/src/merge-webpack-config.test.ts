import { describe, expect, it } from 'vitest';

import type { Configuration } from 'webpack';

import { mergeConfigs } from './merge-webpack-config';

const config: Configuration = {
  devtool: 'source-map',
  entry: {
    bundle: 'index.js',
  },
  output: {
    filename: '[name].js',
  },
  module: {
    rules: [{ use: 'r1' }, { use: 'r2' }],
  },
  // @ts-expect-errors For snapshot readability purposes `plugins` attribute doesn't match the correct type
  plugins: ['p1', 'p2'],
  resolve: {
    enforceExtension: true,
    extensions: ['.js', '.json'],
    alias: {
      A1: 'src/B1',
      A2: 'src/B2',
    },
  },
  optimization: {
    splitChunks: {
      chunks: 'all',
    },
    runtimeChunk: true,
  },
};

describe('mergeConfigs', () => {
  it('merges two full configs in one', () => {
    const customConfig: Configuration = {
      profile: true,
      entry: {
        bundle: 'should_not_be_merged.js',
      },
      output: {
        filename: 'should_not_be_merged.js',
      },
      module: {
        noParse: /jquery|lodash/,
        rules: [{ use: 'r3' }, { use: 'r4' }],
      },
      // @ts-expect-errors For snapshot readability purposes `plugins` attribute doesn't match the correct type
      plugins: ['p3', 'p4'],
      resolve: {
        enforceExtension: false,
        extensions: ['.ts', '.tsx'],
        alias: {
          A3: 'src/B3',
          A4: 'src/B4',
        },
      },
      optimization: {
        // @ts-expect-errors For snapshot readability purposes `minimizer` attribute doesn't match the correct type
        minimizer: ['banana'],
      },
    };

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });

  it('merges partial custom config', () => {
    const customConfig: Configuration = {
      // @ts-expect-errors For snapshot readability purposes `plugins` attribute doesn't match the correct type
      plugins: ['p3'],
      resolve: {
        extensions: ['.ts', '.tsx'],
      },
    };

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });

  it('merges successfully if custom config is empty', () => {
    const customConfig = {};

    const result = mergeConfigs(config, customConfig);

    expect(result).toMatchSnapshot();
  });
});
