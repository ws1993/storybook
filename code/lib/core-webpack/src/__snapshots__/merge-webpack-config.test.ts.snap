// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeConfigs > merges partial custom config 1`] = `
{
  "devtool": "source-map",
  "entry": {
    "bundle": "index.js",
  },
  "module": {
    "rules": [
      {
        "use": "r1",
      },
      {
        "use": "r2",
      },
    ],
  },
  "optimization": {
    "runtimeChunk": true,
    "splitChunks": {
      "chunks": "all",
    },
  },
  "output": {
    "filename": "[name].js",
  },
  "plugins": [
    "p1",
    "p2",
    "p3",
  ],
  "resolve": {
    "alias": {
      "A1": "src/B1",
      "A2": "src/B2",
    },
    "enforceExtension": true,
    "extensions": [
      ".js",
      ".json",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`mergeConfigs > merges successfully if custom config is empty 1`] = `
{
  "devtool": "source-map",
  "entry": {
    "bundle": "index.js",
  },
  "module": {
    "rules": [
      {
        "use": "r1",
      },
      {
        "use": "r2",
      },
    ],
  },
  "optimization": {
    "runtimeChunk": true,
    "splitChunks": {
      "chunks": "all",
    },
  },
  "output": {
    "filename": "[name].js",
  },
  "plugins": [
    "p1",
    "p2",
  ],
  "resolve": {
    "alias": {
      "A1": "src/B1",
      "A2": "src/B2",
    },
    "enforceExtension": true,
    "extensions": [
      ".js",
      ".json",
    ],
  },
}
`;

exports[`mergeConfigs > merges two full configs in one 1`] = `
{
  "devtool": "source-map",
  "entry": {
    "bundle": "index.js",
  },
  "module": {
    "noParse": /jquery\\|lodash/,
    "rules": [
      {
        "use": "r1",
      },
      {
        "use": "r2",
      },
      {
        "use": "r3",
      },
      {
        "use": "r4",
      },
    ],
  },
  "optimization": {
    "minimizer": [
      "banana",
    ],
    "runtimeChunk": true,
    "splitChunks": {
      "chunks": "all",
    },
  },
  "output": {
    "filename": "[name].js",
  },
  "plugins": [
    "p1",
    "p2",
    "p3",
    "p4",
  ],
  "profile": true,
  "resolve": {
    "alias": {
      "A1": "src/B1",
      "A2": "src/B2",
      "A3": "src/B3",
      "A4": "src/B4",
    },
    "enforceExtension": false,
    "extensions": [
      ".js",
      ".json",
      ".ts",
      ".tsx",
    ],
  },
}
`;
